src/connection/socket-connection.ts

```diff
@@ -0,0 +1,215 @@
+        `the subscription with id ${id} was not found`
+      )
+    }
+
+    const connection = this.subscriptions.get(id);
+    await connection.close();
+    this.subscriptions.delete(id);
+  }
+
+  /**
+   * Closes the existing connection and cleans up any listeners or subscriptions.
+   */
+  async close(): Promise<void> {
+    clearInterval(this.heartbeatInterval);
+    // clean up all socket event listeners
+    this.socket.removeAllListeners();
+
+    const closePromises = [];
+    for (const [id, subscription] of this.subscriptions) {
+      closePromises.push(subscription.close());
+      this.subscriptions.delete(id);
+    }
+
+    // close all of the associated subscriptions
+    await Promise.all(closePromises);
+
+    // close the socket.
+    this.socket.close();
+  }
+
+  /**
+   * Pong messages are automatically sent in response to ping messages as required by
+   * the websocket spec. So, no need to send explicit pongs.
+   */
+  private pong(): void {
+    this.isAlive = true;
+  }
+
+  /**
+   * Log the error and close the connection.
+   */
+  private async error(error:Error): Promise<void>{
+    log.error(`SocketConnection error, terminating connection`, error);
+    this.socket.terminate();
+    await this.close();
+  }
+
+  /**
+   * Handles a `JSON RPC 2.0` encoded message.
+   */
+  private async message(dataBuffer: Buffer): Promise<void> {
+    const requestData = dataBuffer.toString();
+    if (!requestData) {
+      return this.send(createJsonRpcErrorResponse(
+        uuidv4(),
+        JsonRpcErrorCodes.BadRequest,
+        'request payload required.'
+      ))
+    }
+
+    let jsonRequest: JsonRpcRequest;
+    try {
+      jsonRequest = JSON.parse(requestData);
+    } catch(error) {
+      const errorResponse = createJsonRpcErrorResponse(
+        uuidv4(),
+        JsonRpcErrorCodes.BadRequest,
+        (error as Error).message
+      );
+
+      return this.send(errorResponse);
+    };
+
+    const requestContext = await this.buildRequestContext(jsonRequest);
+    const { jsonRpcResponse } = await jsonRpcApi.handle(jsonRequest, requestContext);
+    if (jsonRpcResponse.error) {
+      requestCounter.inc({ method: jsonRequest.method, error: 1 });
+    } else {
+      requestCounter.inc({
+        method: jsonRequest.method,
+        status: jsonRpcResponse?.result?.reply?.status?.code || 0,
+      });
+    }
+    this.send(jsonRpcResponse);
+  }
+
+  /**
+   * Sends a JSON encoded Buffer through the Websocket. Accepts an error handler, if none is provided the error is logged.
+   */
+  private send(response: JsonRpcResponse | JsonRpcErrorResponse, onError?: (error?: Error) => void): void {
+    if (!onError) {
+      onError = (error):void => {
+        if(error) { log.error('socket send error', error, response); }
+      }
+    }
+
+    this.socket.send(Buffer.from(JSON.stringify(response)), onError);
+  }
+
+  /**
+   * Subscription Handler used to build the context for a `JSON RPC` API call.
```

Comment:
The "used to build the context" took me a while to know what it means.

Can we reword to say:

```suggestion
   * Creates a subscription handler to send messages matching the subscription to the subscriber.
```

Reply:
Reworded to `Creates a subscription handler to send messages matching the subscription requested.`